### DISCUSSION TITLE
Zram and physical swap

### DISCUSSION THREAD

**Discussion by seedboxer** [MichaIng/DietPi#6289] [Q&A] [answered]
looking to use Zram as priority 1 and a Physical swap as lower as Zram seems to fill up swap quickly.

Im looking at creating the following swaps at boot.  

Currently, zram remains persistent but loses the /swapfile on reboot.

How can i do this as tstab gets overwritten at boot?

NAME       TYPE       SIZE USED PRIO
/dev/zram0 partition  640M 6.8M   -2
/swapfile  file      1024M   0B   -3

**Comment by MichaIng** [comment]
A zram swap has always higher priority than a disk swap. The kernel is smart enough to assume this is always wanted 😎.

> How can i do this as tstab gets overwritten at boot?

It does not get overwritten at boot, at least it should not 🤔. In theory if there is some I/O error or the shutdown is unclean, cutting power before changes from write buffer are synced to disk, this may cause small changes to be lost. Verify once again that both are in `/etc/fstab` and `reboot`. If indeed an entry is lost after reboot, try the same but run `sync; sleep 1; reboot` to force a disk sync and give it a little time to finish before rebooting. If that does not help, then there is a larger issues with your system, like general disk write issues or some sort of mount over `/etc` or `/etc/fstab`.

**Comment by seedboxer** [comment]
Thanks got it to work via 

https://github.com/foundObjects/zram-swap

Although the script didn't set zram swap size correctly easy fix, and used dietpi-manager to set normal swap now have

NAME       TYPE       SIZE USED PRIO
/var/swap  file      1000M   0B   -2
/dev/zram0 partition    2G   0B   15

Works as it should and has lz4 compression which is meant to work best for SBC.

Before was using command /boot/dietpi/func/dietpi-set_swapfile <size> <device>

the script seems to have set things up better.

Using

vm.swappiness = 20
vm.vfs_cache_pressure = 50
vm.dirty_ratio = 60
vm.dirty_background_ratio = 2

Pi4B 4GB. Dietpi has made this little thing a beast !

**Reply by MichaIng** [reply]
> has lz4 compression which is meant to work best for SBC

LZ4 is fast but has bad compression ratio. So it depends on how often the disk swap is used whether it is good or not. If ZSTD (level 1) can reduce disk swapping notably, then it should be preferred. It is nearly as fast as LZ4 but has a much better compression ratio. Of course it highly depends on use case/data etc and kernel version surely plays a role as well, but just to give a starting point for evaluation: https://linuxreviews.org/Comparison_of_Compression_Algorithms

What I take from this is that ZSTD generally is the better default/alround compression algorithm for zram/zswap and filesystem compression, and it has a wide range of compression levels where:
- `1` gives speeds close to LZ4/LZO but much better compression ratio, hence should be usually preferred when performance matters more
- `19` gives compression ratio close to LZMA/XZ/7z but much better (compression) performance, hence should be usually preferred when compression ratio matters more

Btw, why do you use zram-swap if you use a disk swap as well? In this case I'd usually use zswap instead, which does a few things a little smarter. Since zswap is a buffer for other swap spaces, it does not work on its own, so if no disk swap is wanted at all, then zram-swap is required. This is the reason why DietPi allows to create only one swap space, either as disk file or zram, but not both. If both is wanted, then we assume that zswap is the better companion to the disk swap space.

**Reply by seedboxer** [reply]
Thanks for the great info, I have the 2nd swap just as a fail safe really, i would hope its not used as the idea is to keep writes to the disk down.

 A lot of people say to use a 100 Plus vm.swapiness and this doesn't work well I find for a Pi as the swap gets full quick. So I choose to have it just as a buffer. My settings seem fine now and i have changed to the compression you recommend and see how it goes.

I got this little box doing a lot all thanks to Dietpi, it has really opened my eyes to linux as a whole, thanks so much for that.

**Reply by MichaIng** [reply, marked as answer]
> A lot of people say to use a 100 Plus vm.swapiness

Sounds unreasonable to me indeed. Any swap usage slows down the system so I'd always recommend `1` as lowest possible number. Some say that if rarely used data is swapped, there is more space for filesystem cache to speed up file reads for those cached ones. But I think the chance that some files which are read a few times during some process are needed later is often lower than the chance that data from a process which is hold in RAM are needed later. And often performance is not the only relevant aspect but disk wear as well.

As you have zram + disk swap working now, not much need to change it. But for completeness, if you want to test zswap: https://wiki.archlinux.org/title/Zswap
It's Arch Linux docs but works the same way on all Linux distros. On RPi, the kernel boot parameters can be set in `/boot/cmdline.txt` to make it persistent.

Both can be enabled concurrently, but I'm not sure whether this has any benefit or only downsides or whether the kernel is smart enough to use only one of both. It's discussed here, but no final answer:
- https://www.reddit.com/r/linuxquestions/comments/ju4bft/zram_zswap_or_both/
- https://unix.stackexchange.com/questions/554733/effect-of-enabling-zswap-on-zram-without-any-disk-swap/578900

**Comment by andiohn** [comment]
Hey All, sorry to dig this up; I can't figure out how to do disk and zram swap together. I used the `/boot/dietpi/func/dietpi-set_swapfile 1 zram` command but I can't figure out how then then add a lower priority disk swap as well. Can anyone help?

**Reply by Joulinar** [reply]
our tools don't support this scenario. But you could setup it manually yourself. You could try to check with ChatGPT for some instructions.

**Reply by seedboxer** [reply]
You can just create a normal disk swap via Debian command line and just set it to a lower priority. But maybe not worth it unless u have really low amounts of ram. Last time I did it was when I had a Pi 3 which had 1GB  ram. Or you could achieve the same thing with zswap instead I guess, as its a combo or ram and disk swap. Its prob the way id go if I definitely wanted to use disk for swap.

**Reply by MichaIng** [reply]
If you want a higher priority zram based swap space in front of a disk swap space, use "zswap" instead. I.e. configure a regular swap file via `dietpi-set_swapfile`, then enable the zswap kernel feature.

Check whether it is available and enabled already:
```sh
grep -r . /sys/module/zswap/parameters/
```

To enable at runtime:
```sh
echo 1 > /sys/module/zswap/parameters/enabled
```

To enable it permanently, add `zswap.enabled=1` to the kernel command-line parameters, i.e. `/boot/cmdine.txt` in case of RPi.

The zram swap space only makes sense if you do not want to use a disk swap space at all, but extend your RAM space a bit at cost of CPU utilization only. See also the accepted answer above.

**Reply by andiohn** [reply]
Wow! Ya, I already had it enabled, so technically I was using it when I set a 2GB swap on the disk. I'll have to make sure it's working but wow, thank you all for being so fast at replying! @Joulinar, I tried it with chatgpt and it went off on a tangent lol. @seedboxer ya, I worked a bit on that, but thanks for the quick reply. @MichaIng thank you sir! I'm going to try to just use the normal tool and see if zswap is still enabled!

**Comment by andiohn** [comment]
Ya, it's working well, so ya; I just needed to see the status. It would be cool to have the newer better compression options and the alloc as I think it would improve more, but it appears to be working well. Thanks again all.

**Reply by MichaIng** [reply]
See what I said here: https://github.com/MichaIng/DietPi/issues/6205#issuecomment-2828494957
All common compression algorithms are available: zstd, lzo, lz4, lz4hc, deflate, 842

**Comment by amibumping** [comment]
Just a questions about this @MichaIng, in what disk is created zram space? If I had diepit installed on an sd and I had another ssd connected, can I choose where to mount it?

Thank you!

**Reply by MichaIng** [reply]
It is z**ram** space, so of course it is stored in RAM 😉. This space is not mounted anywhere, but accessed via `/dev/zram0` device node.
